sdk/iot/pnp/tests/cmocka/test_az_iot_pnp_client.c

```diff
@@ -0,0 +1,284 @@
+
+static const az_span test_device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID_STR);
+static const az_span test_device_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_HOSTNAME_STR);
+static const az_span test_root_interface_name = AZ_SPAN_LITERAL_FROM_STR(TEST_ROOT_INTERFACE_NAME);
+static const az_span test_user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT);
+static const az_span test_content_type = AZ_SPAN_LITERAL_FROM_STR(TEST_CONTENT_TYPE);
+static const az_span test_content_encoding = AZ_SPAN_LITERAL_FROM_STR(TEST_CONTENT_ENCODING);
+
+static const char test_correct_pnp_user_name[]
+    = "myiothub.azure-devices.net/my_device/"
+      "?api-version=2018-06-30&digital-twin-model-id=" TEST_ROOT_INTERFACE_NAME;
+
+static const char test_correct_pnp_user_name_with_user_agent[]
+    = "myiothub.azure-devices.net/my_device/?api-version=2018-06-30&" TEST_USER_AGENT
+      "&digital-twin-model-id=" TEST_ROOT_INTERFACE_NAME;
+
+#ifndef NO_PRECONDITION_CHECKING
+
+enable_precondition_check_tests()
+
+    static void test_az_iot_pnp_client_init_NULL_client_fails(void** state)
+{
+  (void)state;
+
+  assert_precondition_checked(az_iot_pnp_client_init(
+      NULL, test_device_hostname, test_device_id, test_root_interface_name, NULL));
+}
+
+static void test_az_iot_pnp_client_init_empty_iot_hub_hostname_fails(void** state)
+{
+  (void)state;
+
+  az_iot_pnp_client client;
+  assert_precondition_checked(az_iot_pnp_client_init(
+      &client, AZ_SPAN_NULL, test_device_id, test_root_interface_name, NULL));
+}
+
+static void test_az_iot_pnp_client_init_empty_device_id_fails(void** state)
+{
+  (void)state;
+
+  az_iot_pnp_client client;
+  assert_precondition_checked(az_iot_pnp_client_init(
+      &client, test_device_hostname, AZ_SPAN_NULL, test_root_interface_name, NULL));
+}
+
+static void test_az_iot_pnp_client_init_empty_root_interface_name_fails(void** state)
+{
+  (void)state;
+
+  az_iot_pnp_client client;
+  assert_precondition_checked(
+      az_iot_pnp_client_init(&client, test_device_hostname, test_device_id, AZ_SPAN_NULL, NULL));
+}
+
+#endif // NO_PRECONDITION_CHECKING
+
+static void test_az_iot_pnp_client_get_default_options_succeed(void** state)
+{
+  (void)state;
+
+  az_iot_pnp_client_options options = az_iot_pnp_client_options_default();
+  assert_true(az_span_is_content_equal(options.user_agent, AZ_SPAN_NULL));
+  assert_true(az_span_is_content_equal(options.content_encoding, AZ_SPAN_NULL));
+  assert_true(az_span_is_content_equal(options.content_type, AZ_SPAN_NULL));
+}
+
+static void test_az_iot_pnp_client_init_succeed(void** state)
+{
+  (void)state;
+
+  az_iot_pnp_client client;
+  assert_int_equal(
+      az_iot_pnp_client_init(
+          &client, test_device_hostname, test_device_id, test_root_interface_name, NULL),
+      AZ_OK);
+
+  assert_non_null(az_span_ptr(client._internal.iot_hub_client._internal.iot_hub_hostname));
+  assert_string_equal(TEST_ROOT_INTERFACE_NAME, az_span_ptr(client._internal.root_interface_name));
```

Comment:
`assert_string_equal` shouldn't be used with any of our client members due to the fact that we don't guarantee null termination. A span doesn't guarantee null termination and if someone initializes the client with a non-null terminated span then they can have issues here. [I have a PR here](https://github.com/Azure/azure-sdk-for-c/pull/514) which changes that for other tests.
TLDR: you can use `assert_memory_equal()` here instead.

Reply:
elsewhere as well. there probably shouldn't be any `assert_string_equal` in this whole test file.

Reply:
Agreed - the as_string() tests aren't needed; just in my memcmp() tests.  Though I'm going to check the length of span is also set right just in case.  (Shouldn't matter if memory is equal, but I want to be safe.)  Telemetry tests could potentially add this check, too...

Reply:
yup funny enough exactly what I'm doing now. Modifying all the other tests to ensure:
1. Length of the span is correct.
2. Span capacity is the untouched
3. Content of the span is as it should
4. The span wasn't written past its length (ie set buffer to all `0xFF` and makes sure index of `length` is `0xFF`)

Reply:
My next question was on overflow!, happy you're on it.

A few thoughts:
* First, is there a tool that can help here so we can just get this either statically or run-time a la Valgrind.  (Valgrind I don't think will work since it ties into heap, not stack.  https://stackoverflow.com/questions/5813444/how-to-debug-stack-overwriting-errors-with-valgrind though leads to https://github.com/google/sanitizers/wiki/AddressSanitizer).

* Or if we want to be able to run on arbitrary setup with no dependencies, we should think about building up the 0xFF thingy as generic.  I'm totally hand-wavy here ,but something where you told the size of buffer and it created the bufferLen+1, put the 0xFF in end, but then when you did a span off it ignored the 0xFF as a valid byte?

Reply:
Well Chris mentioned having a generic number to use instead of having it based on the length of the bytes we are expecting. So in my update the spans are all 128.
Essentially I'm updating all the tests to have this at the "setup faze"
```c
  uint8_t test_span_buf[TEST_SPAN_BUFFER_SIZE];
  memset(test_span_buf, 0xFF, _az_COUNTOF(test_span_buf));
  az_span test_span = az_span_init(test_span_buf, 0, _az_COUNTOF(test_span_buf));
```
and then this for check:
```c
  assert_memory_equal(
      test_correct_user_name_with_module_id,
      az_span_ptr(test_span),
      sizeof(test_correct_user_name_with_module_id) - 1);
  assert_int_equal(
      az_span_length(test_span), _az_COUNTOF(test_correct_user_name_with_module_id) - 1);
  assert_int_equal(az_span_capacity(test_span), TEST_SPAN_BUFFER_SIZE);
  assert_int_equal(test_span_buf[az_span_length(test_span)], 0xFF);
```